ledger/internal/evalprefetcher.go

```diff
@@ -0,0 +1,520 @@
+// syncronized extraction
+type preloaderTaskQueue struct {
+	next               *preloaderTaskQueue
+	used               int
+	entries            []*preloaderTask
+	baseIdx            int
+	maxTxnGroupEntries int
+}
+
+func allocPreloaderQueue(count int, maxTxnGroupEntries int) preloaderTaskQueue {
+	return preloaderTaskQueue{
+		entries:            make([]*preloaderTask, count*2+maxTxnGroupEntries*2),
+		maxTxnGroupEntries: maxTxnGroupEntries,
+	}
+}
+
+// enqueue places the queued entry on the queue, returning the latest queue
+// ( in case the current "page" ran out of space )
+func (pq *preloaderTaskQueue) enqueue(t *preloaderTask) {
+	pq.entries[pq.used] = t
+	pq.used++
+	return
+}
+
+func (pq *preloaderTaskQueue) expand() *preloaderTaskQueue {
+	if cap(pq.entries)-pq.used < pq.maxTxnGroupEntries {
+		pq.next = &preloaderTaskQueue{
+			entries: make([]*preloaderTask, cap(pq.entries)*2),
+			used:    0,
+			baseIdx: pq.baseIdx + pq.used,
+		}
+		return pq.next
+	}
+	return pq
+}
+
+func (pq *preloaderTaskQueue) getTaskAtIndex(idx int) (*preloaderTaskQueue, *preloaderTask) {
+	localIdx := idx - pq.baseIdx
+	if pq.used > localIdx {
+		return pq, pq.entries[localIdx]
+	}
+	if pq.next != nil {
+		return pq.next.getTaskAtIndex(idx)
+	}
+	return pq, nil
+}
+
+type accountCreatableKey struct {
+	address basics.Address
+	cidx    basics.CreatableIndex
+}
+
+func loadAccountsAddAccountTask(addr *basics.Address, wg *groupTask, accountTasks map[basics.Address]*preloaderTask, queue *preloaderTaskQueue) {
+	if addr.IsZero() {
+		return
+	}
+	if task, have := accountTasks[*addr]; !have {
+		task := &preloaderTask{
+			address:      addr,
+			groups:       make([]*groupTask, 1, 4),
+			groupIndices: make([]int, 1, 4),
+		}
+		task.groups[0] = wg
+		task.groupIndices[0] = wg.balancesCount
+
+		accountTasks[*addr] = task
+		queue.enqueue(task)
+	} else {
+		task.groups = append(task.groups, wg)
+		task.groupIndices = append(task.groupIndices, wg.balancesCount)
+	}
+	wg.balancesCount++
+}
+
+func loadAccountsAddResourceTask(addr *basics.Address, cidx basics.CreatableIndex, ctype basics.CreatableType, wg *groupTask, resourceTasks map[accountCreatableKey]*preloaderTask, queue *preloaderTaskQueue) {
+	if cidx == 0 {
+		return
+	}
+	key := accountCreatableKey{
+		cidx: cidx,
+	}
+	if addr != nil {
+		key.address = *addr
+	}
+	if task, have := resourceTasks[key]; !have {
+		task := &preloaderTask{
+			address:        addr,
+			groups:         make([]*groupTask, 1, 4),
+			groupIndices:   make([]int, 1, 4),
+			creatableIndex: cidx,
+			creatableType:  ctype,
+		}
+		task.groups[0] = wg
+		task.groupIndices[0] = wg.resourcesCount
+
+		resourceTasks[key] = task
+		queue.enqueue(task)
+	} else {
+		task.groups = append(task.groups, wg)
+		task.groupIndices = append(task.groupIndices, wg.resourcesCount)
+	}
+	wg.resourcesCount++
+}
+
+func (p *accountPrefetcher) prefetch(ctx context.Context) {
```

Comment:
I think a comment explaining the idea of spitting input data into parallel task groups and and figuriting out their completion would be very helpful

Reply:
sure.